scalyr_agent/builtin_monitors/linux_process_metrics.py

```diff
@@ -619,163 +732,214 @@ class ProcessMonitor(ScalyrMonitor):
+
+        # if no processes are running, there is no reason to report the running metric data
+        if not running_pids:
+            self.running_total_metrics = {}
 
     def gather_sample(self):
-        """Collect the per-process metrics for the monitored process.
+        """Collect the per-process tracker for the monitored process(es).
 
-        If the process is no longer running, then attempts to match a new one.
+        For multiple processes, there are a few cases we need to poll for:
+
+        For `n` processes, if there are < n running processes at any given point,
+        we should see if there are any pids or matching expression that gives pids
+        that is currently not running, if so, get its tracker and run it.
+        It is possible that one or more processes ran its course, but we can't assume that
+        and we should keep polling for it.
         """
-        if self.__pid is not None and not self.__is_running():
-            self.__set_process(None)
 
-        if self.__pid is None:
-            self.__set_process(self.__select_process())
+        trackers = []
+        for _pid in list(self.__select_processes()):
+            trackers.append(ProcessTracker(_pid, self._logger, self.__id))
+
+        self._reset_absolute_metrics()
 
-        for gather in self.__gathers:
-            gather.run_single_cycle()
+        for _tracker in trackers:
+            _tracker.set_gathers()
+            _metrics = _tracker.collect()
+            self.record_metrics(_tracker.pid, _metrics)
 
-    def __select_process(self):
-        """Returns the proces id of a running process that fulfills the match criteria.
+        running_pids = [x.pid for x in trackers]
+        self._calculate_running_total(running_pids)
+        self._remove_dead_processes(running_pids)
+
+        self.print_metrics()
+
+    def print_metrics(self):
+        # For backward compatibility, we also publish the monitor id as 'app' in all reported stats.  The old
+        # Java agent did this and it is important to some dashboards.
+
+        for _metric, _metric_value in self.running_total_metrics.items():
+            extra = {'app': self.__id}
+            if _metric.type:
+                extra['type'] = _metric.type
+            self._logger.emit_value(_metric.name, _metric_value, extra)
+
+    def __select_processes(self):
+        """Returns a set of the process ids of processes that fulfills the match criteria.
 
         This will either use the commandline matcher or the target pid to find the process.
-        If no process is matched, None is returned.
+        If no process is matched, an empty list is returned.
 
-        @return: The process id of the matching process, or None
-        @rtype: int or None
+        @return: The set of process id(s) of the matching process, or set()
+        @rtype: set()
         """
+
         sub_proc = None
-        if self.__commandline_matcher is not None:
+
+        if self.__commandline_matcher:
```

Comment:
Also, we should only actually only look for new processes under certain conditions:
* We do not currently have a process we matched
* This is a single process instance and the old process died
* We are in the multiple process case and we are already watching at least one process... and a certain time period has expired (configured via another option).  Essentially, maybe only once every 2 minutes will we look to see if there are more processes we should be monitoring.